### PR TITLE
Table: Add colors block support

### LINKED ISSUE #19659: Improve background color options on table block
### Summary

The table block currently has a hard-coded background color list (themes cannot disable or provide an alternative set of colors).

This was introduced in #10611 as a trade-off—a way to ensure sufficient contrast between background and text colors. Normally providing contrast isn't an issue, the user can change a text color. The table block is a little more complicated since it has a row striping style, alternate rows have a transparent background that the user has no control over, while other rows have the chosen background color with dark text. 

Provided theme colors were not subtle enough to provide contrast, so we went with a set of hard-coded colors that would complement the light background of those alternate rows.

#10611 initially tried dynamically lightening theme colors, but this resulted in those colors looking washed out and dull.

#### Add additional color pickers for table block

Use theme colors like other blocks, allow the user to specify the following colors:

- Primary Background Color. Used for the body of the table.
- Secondary Background Color. Used for Headers, Footers and even rows in the body of the table when row-striping is enabled.
- Text Color

This would also require migrating the current hard-coded colors to custom colors.

#### Add an additional secondary background color picker when row-striping is enabled

Use theme colors like other blocks, and add an extra option for a secondary background color that is only visible when row-striping is enabled. 

A text color option could also be added.

#### Dynamically change text and background colors to provide sufficient contrast

Use the theme colors for the background option (migrate the current color choices to custom colors). When a darker background color is selected dynamically change text color to light. If row-striping is enabled also dynamically change the rows with a transparent background to a black background.

#### Allow the theme to specify an extra range of colors for backgrounds

This could be an option that works in combination with another solution. The theme could provide a set of colors that are suitable for use in backgrounds and provide sufficient contrast with black or white text.

---

Related #16478, #18766

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
## Description
Replace ad-hoc background color selection for tables with colors block support.

- Allows selection of text and background colors
- With recent feature to skip serialization of block support classes/styles, this can now target the table element as required
- Previous color selections are migrated via a deprecation to the appropriate style attribute
- Includes updated fixture tests 

## How has this been tested?
Manually and via fixtures tests using `npm run test-unit` 

#### Test Instructions
1. Create a new post, add a table block and select a background color for the table
2. Save the post
3. Checkout this PR branch, rebuild, then reload the editor
4. Confirm the table block migrated successfully and the previous color selection is still present
5. Select the table block and modify the background and font colors
6. Save post and confirm selections are rendered correctly on the frontend

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
@nosolosw Do you have any thoughts on standardising retrieval of block support provided classes and styles when we are leveraging the "skip serialisation" feature?

With the ability to pick and choose which inner elements we apply these block support props to, it seems like something we'll be doing more of in the near future. In this PR's case, I've leveraged the util from the button block that was extracted from the colors block support hook.

**Comment by oandregal** [comment]
> @nosolosw Do you have any thoughts on standardising retrieval of block support provided classes and styles when we are leveraging the "skip serialisation" feature?

Yeah, we need that extracted to a utility. It seems this util can be something like `__experimentalGetClassesAndStyleForColors`. Somewhere in the `block-editor/hooks` should be a good place. I'd think this is best done in a separate PR before this one.

(Personally, as a reviewer, I welcome many small PRs rather than a big one. Small slots for reviews work best for me, as it allows me to help others while also advance my work without having to block half a day for a single review).

File: packages/block-library/src/table/block.json [hunk 1]
```diff
@@ -125,6 +122,11 @@
 	"supports": {
 		"anchor": true,
 		"align": true,
+		"color": {
+			"__experimentalSkipSerialization": true,
+			"gradients": true,
+			"link": false
```
Comment (oandregal) [inline packages/block-library/src/table/block.json]: Probably not necessary at the moment as the existing `getColorAndStyleProps` doesn't process link color, but it brings the question: shouldn't it?

**Comment by talldan** [comment]
When testing this, will need to make sure the 'stripes' block style is well tested and works as expected. Also worth testing with table headers and footers active.

I did a quick test, and looks like text color doesn't seem to work for 'stripes'.

File: packages/block-library/src/table/block.json (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/block.json]: I believe the intent of the current `getColorAndStyleProps` was to only bring in what was needed for the button block. That happened to line up with what was needed here for the table block. 

More than likely, a generic utility would need to take into account the current block's configuration of block supports.

**Comment by aaronrobertshaw** [comment]
> Yeah, we need that extracted to a utility. It seems this util can be something like __experimentalGetClassesAndStyleForColors. Somewhere in the block-editor/hooks should be a good place. I'd think this is best done in a separate PR before this one.

My thinking was as more block support features are implemented and there are more cases of skipping serialization. Having a utility to retrieve the css classes and inline styles across block support features would be good if possible.

`__experimentalGetBlockSupportClassesAndStyles( [ 'colors', 'border' ] )`

That might be over-the-top though in hindsight.

> (Personally, as a reviewer, I welcome many small PRs rather than a big one. Small slots for reviews work best for me, as it allows me to help others while also advance my work without having to block half a day for a single review).

I'm happy to create a couple of PRs, extracting `getColorAndStyleProps` into a utility in `block-editor/hooks`, one to update the button block accordingly, then updating this as well.

**Comment by aaronrobertshaw** [comment]
> I did a quick test, and looks like text color doesn't seem to work for 'stripes'.

@talldan Any chance you can let me know which theme you were using when you tested? 

The "stripes" block style works ok for me with the custom colors. When you found it not to work, was that related to the arbitrary background color for the striped rows clashing with a selected text color?

TableColors-2

**Comment by talldan** [comment]
@aaronrobertshaw I did some further testing, and I'm seeing that text colors don't work when a background color is selected with the stripes style:
table-colors

Using Twenty Twenty One. I think it's a theme issue, it works ok with editor theme styles disabled in the settings. Is that something you can see as well? If so, it'd be good to follow up with a trac ticket.

**Comment by aaronrobertshaw** [comment]
> Using Twenty Twenty One. I think it's a theme issue, it works ok with editor theme styles disabled in the settings. Is that something you can see as well? If so, it'd be good to follow up with a trac ticket.

From memory, I was likely using the TT1 Blocks theme. At least thats the first explanation that comes to mind as the GIF in my previous comment shows the font colors being applied correctly both in the editor and frontend. 

The standard Twenty Twenty One theme may not be qualifying for the `has-<color>-text-color` classes to be generated. The `getColorAndStyleProps` function currently forces inline styles for the colors in the editor.

I'll test further and see if I can confirm that.

**Comment by aaronrobertshaw** [comment]
After further testing, I found two themes that introduced conflicting styles preventing the selected table colors from displaying correctly. These themes were:

- TwentyTwentyOne
- Maywood

TableColors-ThemeIssue

In other themes, the text color was displayed correctly in both the editor and the frontend. These themes included:

- TT1 Blocks (block based TwentyTwentyOne)
- TwentyTwenty
- TwentyTwenty (block based)
- Q
- Varia
- Seedlet
- Hever

> If so, it'd be good to follow up with a trac ticket.

If the approach to applying block support styles in this PR is accepted I can create any trac tickets needed so themes get updated to honour users' color selections.

**Review by talldan — APPROVED** [review]
Thanks for investigating those issues. Would be great to follow up with trac tickets.

The changes here look good, and users will be very happy to see this shipped. Nice work 🎉

**Comment by talldan** [comment]
To get this merged, looks like it may require a rebase and reformat of the json files as per https://github.com/WordPress/gutenberg/pull/30714.

**Comment by aaronrobertshaw** [comment]
> The changes here look good, and users will be very happy to see this shipped. Nice work 🎉
> To get this merged, looks like it may require a rebase and reformat of the json files as per #30714.

Thanks for all the effort on the reviews 👍 

As part of addressing the earlier suggestion to move the `getColorAndStyleProps` function into a new utility available via the block support hooks, I've rebased and updated this PR to utilise that. https://github.com/WordPress/gutenberg/pull/30869 

My apologies for the updates post approval!

**Comment by talldan** [comment]
> My apologies for the updates post approval!

No worries! I think you could've merged and then followed up with the other change in another PR, but it's ok to do it this way as well.

I'm not all that familiar with the support hook, so would appreciate if @nosolosw could give the thumbs up to that change.

**Comment by aaronrobertshaw** [comment]
Thanks for taking another look at this one 👍 

> I'm not all that familiar with the support hook, so would appreciate if @nosolosw could give the thumbs up to that change.

The [color block support utilities](https://github.com/WordPress/gutenberg/pull/30869) have now been merged as well as [updates to the button block](https://github.com/WordPress/gutenberg/pull/30870) to use them. Hopefully, this should only need a quick rubber stamp from here.

File: packages/block-library/src/table/edit.js [hunk 2]
```diff
@@ -58,29 +57,6 @@ import {
 	isEmptyTableSection,
 } from './state';
 
-const BACKGROUND_COLORS = [
```
Comment (jorgefilipecosta) [inline packages/block-library/src/table/edit.js:61]: There was a requirement that the table block given its specifies had a specific color palette different from the other blocks, instead of using the default or theme colors. This requirement added some complexity and was the reason createCustomColorsHOC was implemented.
This PR removes that mechanism and makes the table block use the same colors as the other blocks. Is this change intentional and the requirement of having specific colors was removed?


I guess we could easily keep the specific colors by adding a table block specific color palette to lib/experimental-default-theme.json. Would that make sense?

File: packages/block-library/src/table/deprecated.js [hunk 3]
```diff
@@ -6,13 +6,261 @@ import classnames from 'classnames';
...
+} from '@wordpress/block-editor';
 
 const supports = {
 	align: true,
 };
 
+// As the previous arbitrary colors won't match theme color palettes, the hex
+// value will be mapped to the style.color.background attribute as if it was
+// a custom color selection.
+const oldColors = {
```
Comment (jorgefilipecosta) [inline packages/block-library/src/table/deprecated.js:22]: We are making every table block that exists use inline styles instead of a color class. But we will still need to keep the old classes so the blocks properly work on the frontend even if they are never opened on the editor.

Could we leave the attributes as they were instead (keeping the classes and named attributes)? I guess we could keep the old colors as the default table block palette on lib/experimental-default-theme.json. Given that the classes still exist and will always be preset I guess the blocks should still look ok without any migration.

File: packages/e2e-tests/fixtures/blocks/core__table__deprecated-1.html [hunk 4]
```diff
@@ -1,3 +1,3 @@
-<!-- wp:core/table -->
-<table class="wp-block-table"><thead><tr><th>Version</th><th>Musician</th><th>Date</th></tr></thead><tbody><tr><td><a href="https://wordpress.org/news/2003/05/wordpress-now-available/">.70</a></td><td>No musician chosen.</td><td>May 27, 2003</td></tr><tr><td><a href="https://wordpress.org/news/2004/01/wordpress-10/">1.0</a></td><td>Miles Davis</td><td>January 3, 2004</td></tr><tr><td>Lots of versions skipped, see <a href="https://codex.wordpress.org/WordPress_Versions">the full list</a></td><td>&hellip;</td><td>&hellip;</td></tr><tr><td><a href="https://wordpress.org/news/2015/12/clifford/">4.4</a></td><td>Clifford Brown</td><td>December 8, 2015</td></tr><tr><td><a href="https://wordpress.org/news/2016/04/coleman/">4.5</a></td><td>Coleman Hawkins</td><td>April 12, 2016</td></tr><tr><td><a href="https://wordpress.org/news/2016/08/pepper/">4.6</a></td><td>Pepper Adams</td><td>August 16, 2016</td></tr><tr><td><a href="https://wordpress.org/news/2016/12/vaughan/">4.7</a></td><td>Sarah Vaughan</td><td>December 6, 2016</td></tr></tbody></table>
-<!-- /wp:core/table -->
+<!-- wp:table {"backgroundColor":"subtle-pale-blue"} -->
```
Comment (jorgefilipecosta) [inline packages/e2e-tests/fixtures/blocks/core__table__deprecated-1.html:1]: What makes us need to change and existing deprecation? Could we leave this deprecation without changes and create a new deprecation specific to the new case we want to test?

File: packages/block-library/src/table/edit.js [hunk 5]
```diff
@@ -499,27 +473,14 @@ function TableEdit( {
...
 					</PanelBody>
-					<PanelColorSettings
-						title={ __( 'Color' ) }
-						initialOpen={ false }
-						colorSettings={ [
-							{
-								value: backgroundColor.color,
-								onChange: setBackgroundColor,
-								label: __( 'Background color' ),
-								disableCustomColors: true,
```
Comment (jorgefilipecosta) [inline packages/block-library/src/table/edit.js:510]: Previously custom colors were intentionally disabled for the table block. This PR enables custom colors for the table block. Is this change intentional? Should we keep them disabled using lib/experimental-default-theme.json?

File: packages/block-library/src/table/edit.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/edit.js:61]: The idea of having a custom color palette for the table only might work. I wonder if it is missing an opportunity though.

The table block was identified as benefiting from the colors block support in https://github.com/WordPress/gutenberg/issues/28913. The colors block support would provide more control for users including the ability to set text color. My understanding was the block support also then includes contrast checking for the background and text colors.

There is also [this issue](https://github.com/WordPress/gutenberg/issues/19659) that discusses the need for additional table color options.

File: packages/block-library/src/table/deprecated.js (hunk 3 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/deprecated.js:22]: > We are making every table block that exists use inline styles instead of a color class. But we will still need to keep the old classes so the blocks properly work on the frontend even if they are never opened on the editor.

If the block's are not opened in the editor and therefore not migrated. They would still keep the old CSS classes. This is why those classes still exist within the `style.css` file. 

> Could we leave the attributes as they were instead (keeping the classes and named attributes)? I guess we could keep the old colors as the default table block palette on lib/experimental-default-theme.json. Given that the classes still exist and will always be preset I guess the blocks should still look ok without any migration.

If the table block is given its own custom color palette, as you suggested, with the old colors under the same names, their CSS classes should be unchanged. With that approach, the original `backgroundColor` attribute could remain as is, instead of migrating it to the `style.color.background` attribute.

File: packages/e2e-tests/fixtures/blocks/core__table__deprecated-1.html (hunk 4 above)
Comment (aaronrobertshaw) [reply packages/e2e-tests/fixtures/blocks/core__table__deprecated-1.html:1]: My understanding is that the ideal order for deprecations is reverse chronological. In keeping with past advice to that effect, I added the deprecation for the colors block support changes to the beginning of the deprecated array.

I regenerated the deprecation fixtures so they still aligned. That is, the first deprecation in the array matches the fixture named `*_deprecated-1.html`.

Deprecations in general feel very counter intuitive. So making the fixtures not align didn't see like an option. Leaving a choice to either:
- Append the new deprecation to the array and only add new fixtures (which went against past recommendations for deprecation order)
- Prepend the new deprecation and regenerate fixtures (the downside being inconvenience when reviewing).

I went with the later as it was a one time cost and got the slight optimisation of the reverse-chronologically order deprecations.

File: packages/block-library/src/table/edit.js (hunk 5 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/edit.js:510]: With the adoption of the colors block support increasing the user's control over colors including font color, it seemed like this was the preferred direction to go. More so when taking into account some of the conversation in https://github.com/WordPress/gutenberg/issues/19659.

**Comment by aaronrobertshaw** [comment]
Thank you for the great feedback @jorgefilipecosta 👍 

It would be great to get some input from others regarding whether we should maintain a custom color palette for the table block as suggested as well as allow custom colors. @talldan and @nosolosw do you have any ideas or preferences on this?

**Comment by talldan** [comment]
I would defer to the conversation on #19659, where it has been made very clear that the current limited palette is a problem for users.

See @kjellr's comment here:
https://github.com/WordPress/gutenberg/issues/19659#issuecomment-794232681

I think this PR would be a good first step, but there are additional challenges that I think it would be good to devote some time to solving:
- Users can't specify different colors for table headers and footers, which is a pretty common thing to do.
- The striped style generally causes contrast issues, since the table has two background colors when this is active, but the user can only specify one of those background colors and one text color. Generally it's easiest to go for a lighter background color, but themes often don't provide a decent range of options.

File: packages/block-editor/src/hooks/use-color-props.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
+/**
```
Comment (oandregal) [inline packages/block-editor/src/hooks/use-color-props.js]: Wasn't this merged as part of https://github.com/WordPress/gutenberg/pull/30869? I guess this needs a rebase.

File: packages/block-library/src/table/block.json [hunk 7]
```diff
@@ -125,6 +122,10 @@
 	"supports": {
 		"anchor": true,
 		"align": true,
+		"color": {
```
Comment (oandregal) [inline packages/block-library/src/table/block.json:125]: I've tested it and the background color works as in `trunk`. That part is ready to land in my view. 

However, I'm not sure about text color for this block: it affects both text and the table borders. Shouldn't those things be configurable separately?

One thing we could do to land the parts of this PR that are ready is this: only enable background (and gradients) for now and prepare a follow-up to explore the relationship between text & border gathering more design feedback.

**Comment by oandregal** [comment]
In terms of how this uses the block supports: it's fine, although it looks like we need to rebase to get the latest utilities (and remove those changes from this PR).

I suggested at https://github.com/WordPress/gutenberg/pull/30791#discussion_r620218821 that we start only with background for this PR (which works as of today, just with more theme colors) and tackle text/border separately.

**Comment by oandregal** [comment]
By the way, I've noticed the dropdown for the default style looks a bit odd (both Firefox and Chrome on Linux):

Captura de ecrã de 2021-04-26 16-22-52

File: packages/block-editor/src/hooks/use-color-props.js (hunk 6 above)
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/use-color-props.js]: ✅ Rebased.

**Comment by aaronrobertshaw** [comment]
> By the way, I've noticed the dropdown for the default style looks a bit odd (both Firefox and Chrome on Linux):

Unfortunately, I don't have a Linux VM handy at the moment to test. It appears OK for me across browsers under Mac and Windows using a few different themes.

Is it just this PR you noticed the issue with and is it still an issue after the rebase?

File: packages/block-library/src/table/block.json (hunk 7 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/block.json:125]: > However, I'm not sure about text color for this block: it affects both text and the table borders. Shouldn't those things be configurable separately?

This comes from the active theme. Several themes set the border color to `initial` which means they get set to `currentcolor` and take on the font color. With the new border color block support it could be configurable separately. Which as you say, is better done in a follow up PR.

I thought the greater concern is ensuring sufficient contrast between the background and text colors. If we adopt the colors block support for the background but not text color, do we just create a bigger issue? Is that acceptable if we are looking to dial back this PR to only update background color?

**Comment by oandregal** [comment]
> Is it just this PR you noticed the issue with and is it still an issue after the rebase?

Oh, sorry, hadn't tested on `trunk`: I just did and it also happens over there.

File: packages/block-library/src/table/block.json (hunk 7 above)
Comment (oandregal) [reply packages/block-library/src/table/block.json:125]: Thought about this more: the current behavior is also text + border, so we should be fine.

I guess what I'm uncomfortable with is that this block has a larger design surface than many other blocks and it doesn't look super great at the moment (two background colors -with stripes-, border & text, etc). These are things that Dan brought up already https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094 so let's merge and iterate.

**Review by oandregal — APPROVED** [review]
Tested this a bit and works as expected (tested deprecations with old colors and are properly migrated to the new system). From the point of view of the block supports, it's also ready to land.

File: packages/block-library/src/table/block.json (hunk 7 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/table/block.json:125]: Sounds like a plan 👍 

I'll merge and take a look at some options to address https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094.

**Comment by aaronrobertshaw** [comment]
I've created https://github.com/WordPress/gutenberg/issues/31261 to track the desired follow ups mentioned in https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094
